### ISSUE TITLE
`<{f16,f32,f64,f128} as Rem>::rem` are not remainder of truncated division, as documented

### ISSUE THREAD

**Issue by traviscross** [rust-lang/rust#133758]
Our documentation for `impl Rem for {f16, f32, f64, f128}` [says](https://doc.rust-lang.org/std/ops/trait.Rem.html#impl-Rem-for-f64):

> The remainder has the same sign as the dividend and is computed as: `x - (x / y).trunc() * y`.

But that's not true.  E.g.:

```rust
fn main() {
    let (x, y) = (11f64, 1.1f64);
    assert_eq!(x - (x / y).trunc() * y, x % y);
    //~^ PANIC
    // assertion `left == right` failed
    // left: 0.0
    // right: 1.0999999999999992
}
```

That is, `x % y` is documented to be the remainder of truncated division, but it's clearly not doing that.  This makes it difficult to correctly encode algorithms that rely on this semantic.

This tripped me up, e.g., when authoring:

- https://github.com/rust-lang/rust/pull/133755

cc #133485 #107904

cc @cuviper @Noratrieb @tczajka @Neutron3529 @BartMassey @rust-lang/libs @rust-lang/libs-api

**Comment by traviscross** [comment]
Notably, and perhaps uncoincidentally, C's `fmod` has the same [documented](https://en.cppreference.com/w/c/numeric/math/fmod) behavior...

> The floating-point remainder of the division operation `x / y` calculated by this function is exactly the value `x - n * y`, where `n` is `x / y` with its fractional part truncated.

...and the same problem:

```rust
unsafe extern "C" {
    safe fn fmod(x: f64, y: f64) -> f64;
}

fn main() {
    let (x, y) = (11f64, 1.1f64);
    assert_eq!(x - (x / y).trunc() * y, fmod(x, y));
    //~^ PANIC
    // assertion `left == right` failed
    // left: 0.0
    // right: 1.0999999999999992
}
```

**Comment by quaternic** [comment]
Indeed, Rust's floating point `%` is actually equivalent to [`fmod`](https://en.cppreference.com/w/cpp/numeric/math/fmod). The documentation is only correct if you interpret the expression `x - (x / y).trunc() * y` as an exact formula. (The wording of "computed as" certainly doesn't help one towards that reading.)

**Comment by traviscross** [comment]
Yes, and the C docs state it even more clearly.  They say "it is *exactly* the value...".

**Comment by traviscross** [comment]
Probably I agree, as a practical matter, it should be clarified in the docs.  Strictly speaking, though, in my reading of the C23 [(pre-)standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3096.pdf), section 7.12.10, either behavior would be acceptable:

> The `fmod` functions return the value `x − ny`, for some integer `n` such that, if `y` is nonzero, the result has the same sign as `x` and magnitude less than the magnitude of `y`. If `y` is zero, whether a domain error occurs or the `fmod` functions return zero is implementation-defined.

That is, the standard is satisfied whether we choose `9` or `10` for `n` here.

**Comment by quaternic** [comment]
Oh, I see. It's specified more loosely than the related `remainder` which is mandated by IEEE754.

So `fmod(11.0, 1.1)` is *allowed* to return either of
- `11.0 - 9.0 * 1.1 == 1.0999999999999996` 
- `11.0 - 10.0 * 1.1 == 0.0`
in addition to the exact result of `1.0999999999999992_f64`.

It's starting to look like a shaky foundation to be basing our operations on.

**Comment by tczajka** [comment]
> That is, the standard is satisfied whether we choose 9 or 10 for n here.

> So, `fmod(11.0, 1.1)` is _allowed_ to return either of
> 
> * `11.0 - 9.0 * 1.1 == 1.0999999999999996`
> * `11.0 - 10.0 * 1.1 == 0.0`
> 
> (in addition to the exact result of `1.0999999999999992_f64`)

No, this is incorrect. The exact value is the only value that is allowed by the C standard (for an implementation that supports IEEE-754 with subnormals), as specified by F.10.7.1:
> When subnormal results are supported, the returned value is exact and is independent of the current
rounding direction mode.

The C expression `11.0 - 10.0  *1.1` is irrelevant. Section 7.12.9 says:
> x − ny

That's a mathematical formula, not a C expression (as is clear by the lack of a `*` operator between n and y). For n=10 the value of the expression is negative which is not allowed.

**Comment by traviscross** [comment]
Agreeing and extending, IEEE 754-2008 says in 5.3.1:

> When `y ≠ 0`, the remainder `r = remainder(x, y)` is defined for finite `x` and `y` regardless of the rounding-direction attribute by the mathematical relation `r = x − y × n` , where n is the integer nearest the exact number `x/y` ; whenever `| n − x/y | = ½` , then `n` is even. Thus, the remainder is always exact. If `r = 0`, its sign shall be that of `x`. `remainder(x, ∞)` is `x` for finite `x`.

Then C23 F.10.7.1 says:

> The `double` version of `fmod` behaves as though implemented by:
>
> ```c
> #include <math.h>
> #include <fenv.h>
> #pragma STDC FENV_ACCESS ON
> double fmod(double x, double y)
> {
>     double result;
>     result = remainder(fabs(x), (y = fabs(y)));
>     if (signbit(result)) result += y;
>     return copysign(result, x);
> }
> ```

**Comment by 11happy** [comment]
Hello @traviscross , I am interested to work on this issue to fix `div_euclid` could you please assign this to me. 
Thank you

**Comment by tczajka** [comment]
This is a duplicate of #107904.

**Comment by traviscross** [comment]
Actually, I don't think this is a duplicate of that.  This is a separate documentation bug.  We just need to say explicitly in the docs that the stated calculation, `x - (x / y).trunc() * y`, is done in infinite precision rather than literally, which is how it currently reads and what the example given strongly suggests.

**Comment by tczajka** [comment]
OK but then title of the ticket should be clarified. `rem` **is** in fact the remainder of truncated division.
